<a href="https://colab.research.google.com/github/graphy-young/algorithm/blob/main/Baekjoon_Online_Judge.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1단계. 입출력과 사칙연산

입력, 출력과 사칙연산을 연습해 봅시다. Hello World!

## Q2557. Hello World

### 문제
    Hello World!를 출력하시오.

### 입력
    없음

### 출력
    Hello World!를 출력하시오.

### Code

In [ ]:
print('Hello World!')

Hello World!


## Q10718. We love kriii

### 문제
    ACM-ICPC 인터넷 예선, Regional, 그리고 World Finals까지 이미 2회씩 진출해버린 kriii는 미련을 버리지 못하고 왠지 모르게 올 해에도 파주 World Finals 준비 캠프에 참여했다.
    
    대회를 뜰 줄 모르는 지박령 kriii를 위해서 격려의 문구를 출력해주자.

### 입력
    본 문제는 입력이 없다.

### 출력
    두 줄에 걸쳐 "강한친구 대한육군"을 한 줄에 한 번씩 출력한다.

### Code

In [ ]:
print('강한친구 대한육군\n강한친구 대한육군')

강한친구 대한육군
강한친구 대한육군


In [ ]:
for i in range(2): print('강한친구 대한육군')

강한친구 대한육군
강한친구 대한육군


## Q10171. 고양이

### 문제
    아래 예제와 같이 고양이를 출력하시오.

### 입력
    없음.

### 출력
    고양이를 출력한다.

### Code

In [ ]:
print('''\    /\ 
 )  ( ')
(  /  )
 \(__)|''')

\    /\ 
 )  ( ')
(  /  )
 \(__)|


## Q10172. 개

### 문제
    아래 예제와 같이 개를 출력하시오.

### 입력
    없음.

### 출력
    개를 출력한다.

### Code

In [ ]:
print('''|\_/|
|q p|   /}
( 0 )"""\\
|"^"`    |
||_/=\\\\__|''')

|\_/|
|q p|   /}
( 0 )"""\
|"^"`    |
||_/=\\__|


## Q1000. A+B

### 문제
    두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 A와 B가 주어진다. (0 < A, B < 10)

### 출력
    첫째 줄에 A+B를 출력한다.

### Code

In [ ]:
func = lambda x: sum(x)

values = input().split()

for idx in range(len(values)): values[idx] = int(values[idx])
print(func(values))

1 2
3


## Q1001. A-B

### 문제
    두 정수 A와 B를 입력받은 다음, A-B를 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 A와 B가 주어진다. (0 < A, B < 10)

### 출력
    첫째 줄에 A-B를 출력한다.

### 예제 입력 1
    3 2
### 예제 출력 1
    1

### Code

In [ ]:
def func(*args):
    values = args[0]
    for idx in range(len(values)):
        if idx == len(values)-1: return values[0]
        reverse_idx = (len(values)-1)-idx
        values[0] -= values[reverse_idx]

values = input().split()

for idx in range(len(values)): values[idx] = int(values[idx])

print(func(values))

3 2
1


## Q10998. AxB

### 문제
    두 정수 A와 B를 입력받은 다음, A×B를 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 A와 B가 주어진다. (0 < A, B < 10)

### 출력
    첫째 줄에 A×B를 출력한다.

### 예제 입력 1
    1 2
### 예제 출력 1
    2

### 예제 입력 2
    3 4
### 예제 출력 2
    12

### Code

In [ ]:
print((lambda x: int(x[0]) * int(x[2]))(input()))

1 2


2

## Q. 1008. A/B

### 문제
    두 정수 A와 B를 입력받은 다음, A/B를 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 A와 B가 주어진다. (0 < A, B < 10)

### 출력
    첫째 줄에 A/B를 출력한다. 실제 정답과 출력값의 절대오차 또는 상대오차가 10-9 이하이면 정답이다.

### 예제 입력 1
    1 3
### 예제 출력 1
    0.33333333333333333333333333333333

**10-9 이하의 오차를 허용한다는 말은 꼭 소수 9번째 자리까지만 출력하라는 뜻이 아니다.**

### 예제 입력 2
    4 5
### 예제 출력 2
    0.8

### Code

In [ ]:
print( (lambda x: int(x[0]) / int(x[2]))(input()) )

1 3
0.3333333333333333


## Q10869. 사칙연산

### 문제
    두 자연수 A와 B가 주어진다. 이때, A+B, A-B, A*B, A/B(몫), A%B(나머지)를 출력하는 프로그램을 작성하시오. 

### 입력
    두 자연수 A와 B가 주어진다. (1 ≤ A, B ≤ 10,000)

### 출력
    첫째 줄에 A+B, 둘째 줄에 A-B, 셋째 줄에 A*B, 넷째 줄에 A/B, 다섯째 줄에 A%B를 출력한다.

### 예제 입력 1
    7 3
### 예제 출력 1
    10
    4
    21
    2
    1

### Code

In [ ]:
def func(args):
    values = [int(x) for x in args.split()]
    a, b = values
    if (1 <= a <= 10000) and (1 <= b <= 10000): 
        print(a+b)
        print(a-b)
        print(a*b)
        print(a//b)
        print(a%b)
    else: print('Worng input')

func(input())

## Q10430. 나머지

### 문제
    (A+B)%C는 (A%C + B%C)%C 와 같을까?
    
    (A×B)%C는 (A%C × B%C)%C 와 같을까?
    
    세 수 A, B, C가 주어졌을 때, 위의 네 가지 값을 구하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 A, B, C가 순서대로 주어진다. (2 ≤ A, B, C ≤ 10000)

### 출력
    첫째 줄에 (A+B)%C, 둘째 줄에 (A%C + B%C)%C, 셋째 줄에 (A×B)%C, 넷째 줄에 (A%C × B%C)%C를 출력한다.

### 예제 입력 1
    5 8 4
### 예제 출력 1
    1
    1
    0
    0

### Code

In [ ]:
def func(args):
    values = [int(x) for x in args.split()]
    for val in values:
        if 2 <= val <= 10000: pass
        else: raise ValueError
    a, b, c = values[0], values[1], values[2]
    print( (a+b)%c )
    print( (a%c + b%c)%c )
    print( (a*b)%c )
    print( (a%c * b%c)%c )

func(input())

5 8 4
1
1
0
0


## Q2588. 곱셈


### 문제
    (세 자리 수) × (세 자리 수)는 다음과 같은 과정을 통하여 이루어진다.
    
    (1)과 (2)위치에 들어갈 세 자리 자연수가 주어질 때 (3), (4), (5), (6)위치에 들어갈 값을 구하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 (1)의 위치에 들어갈 세 자리 자연수가, 둘째 줄에 (2)의 위치에 들어갈 세자리 자연수가 주어진다.

### 출력
    첫째 줄부터 넷째 줄까지 차례대로 (3), (4), (5), (6)에 들어갈 값을 출력한다.

### 예제 입력 1
    472
    385
### 예제 출력 1
    2360
    3776
    1416
    181720

### Code

In [ ]:
def func():
    values = list()
    for i in range(2): values.append(int(input()))
    #values = [int(x) for x in args.split('\n')]
    multiplicand = values[0] # 피승수
    multiplier = values[1] # 승수
    for i in range( len(str(multiplier))-1, -1, -1 ):
        print(multiplicand * int(str(multiplier)[i]))
    print(multiplicand * multiplier)
        
func()

472
385
2360
3776
1416
181720


# 2단계. if문

if문을 사용해 봅시다.

## 1330. 두 수 비교하기

### 문제
    두 정수 A와 B가 주어졌을 때, A와 B를 비교하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 A와 B가 주어진다. A와 B는 공백 한 칸으로 구분되어져 있다.

### 출력
    첫째 줄에 다음 세 가지 중 하나를 출력한다.
    A가 B보다 큰 경우에는 '>'를 출력한다.
    A가 B보다 작은 경우에는 '<'를 출력한다.
    A와 B가 같은 경우에는 '=='를 출력한다.

### 제한
    -10,000 ≤ A, B ≤ 10,000

### 예제 입력 1
    1 2
### 예제 출력 1
    <
### 예제 입력 2
    10 2
### 예제 출력 2
    >
### 예제 입력 3
    5 5
### 예제 출력 3
    ==

### Code

In [ ]:
def comparison(args):
    values = [int(x) for x in args.split()]
    for val in values:
        if -10000 <= val <= 10000: pass
        else: raise ValueError
    a, b = values
    if a > b: print('>')
    elif a < b: print('<')
    else: print('==')

comparison(input())

## 9498. 시험 성적

### 문제
    시험 점수를 입력받아 90 ~ 100점은 A, 80 ~ 89점은 B, 70 ~ 79점은 C, 60 ~ 69점은 D, 나머지 점수는 F를 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 시험 점수가 주어진다. 시험 점수는 0보다 크거나 같고, 100보다 작거나 같은 정수이다.

### 출력
    시험 성적을 출력한다.

### 예제 입력 1
    100
### 예제 출력 1
    A

### Code

In [ ]:
def testGrade(arg):
    score = int(arg)
    if 90 <= score <= 100: 
        print('A')
    elif 80 <= score <= 89: 
        print('B')
    elif 70 <= score <= 79:
        print('C')
    elif 60 <= score <= 69:
        print('D')
    else: 
        print('F')
        
testGrade(input())

## 2753. 윤년

### 문제
    연도가 주어졌을 때, 윤년이면 1, 아니면 0을 출력하는 프로그램을 작성하시오.
    
    윤년은 연도가 4의 배수이면서, 100의 배수가 아닐 때 또는 400의 배수일 때 이다.
    
    예를들어, 2012년은 4의 배수라서 윤년이지만, 1900년은 4의 배수이지만, 100의 배수이기 때문에 윤년이 아니다.
    
    하지만, 2000년은 400의 배수이기 때문에 윤년이다.

### 입력
    첫째 줄에 연도가 주어진다. 연도는 1보다 크거나 같고, 4000보다 작거나 같은 자연수이다.

### 출력
    첫째 줄에 윤년이면 1, 아니면 0을 출력한다.

### 예제 입력 1
    2000
### 예제 출력 1
    1

### Code

In [ ]:
def isLeafYear(arg):
    year = int(arg)
    if 1 <= year <= 4000: 
        pass
    else: 
        raise ValueError
    if (year%4 == 0) and ( (year%100 != 0) or (year%400 == 0) ):
        print(1)
    else: 
        print(0)
        
isLeafYear(input())

## 10817. 세 수 (삭제됨)

### 문제
    세 정수 A, B, C가 주어진다. 이때, 두 번째로 큰 정수를 출력하는 프로그램을 작성하시오. 

### 입력
    첫째 줄에 세 정수 A, B, C가 공백으로 구분되어 주어진다. (1 ≤ A, B, C ≤ 100)

### 출력
    두 번째로 큰 정수를 출력한다.

### 예제 입력 1
    20 30 10
### 예제 출력 1
    20
### 예제 입력 2
    30 30 10
### 예제 출력 2
    30
### 예제 입력 3
    40 40 40
### 예제 출력 3
    40
### 예제 입력 4
    20 10 10
### 예제 출력 4
    10

### Code

In [ ]:
def threeDigits(args):
    values = [int(x) for x in args.split()]
    for val in values:
        if 1 <= val <= 100: 
            pass
        else: 
            raise ValueError
    values.sort()
    print(values[1])
    
threeDigits(input())

## 14681. 사분면 고르기

### 문제
    흔한 수학 문제 중 하나는 주어진 점이 어느 사분면에 속하는지 알아내는 것이다. 사분면은 아래 그림처럼 1부터 4까지 번호를 갖는다. "Quadrant n"은 "제n사분면"이라는 뜻이다.

    예를 들어, 좌표가 (12, 5)인 점 A는 x좌표와 y좌표가 모두 양수이므로 제1사분면에 속한다. 점 B는 x좌표가 음수이고 y좌표가 양수이므로 제2사분면에 속한다.

    점의 좌표를 입력받아 그 점이 어느 사분면에 속하는지 알아내는 프로그램을 작성하시오. 단, x좌표와 y좌표는 모두 양수나 음수라고 가정한다.

### 입력
    첫 줄에는 정수 x가 주어진다. (−1000 ≤ x ≤ 1000; x ≠ 0) 다음 줄에는 정수 y가 주어진다. (−1000 ≤ y ≤ 1000; y ≠ 0)

### 출력
    점 (x, y)의 사분면 번호(1, 2, 3, 4 중 하나)를 출력한다.

### 예제 입력 1 
    12
    5
### 예제 출력 1 
    1
### 예제 입력 2 
    9
    -13
### 예제 출력 2 
    4

### Code

In [ ]:
points = dict()
points['x'] = int(input())
points['y'] = int(input())

if points['x'] > 0 and points['y'] > 0:
  print(1)
elif points['x'] < 0 and points['y'] > 0:
  print(2)
elif points['x'] > 0 and points['y'] < 0:
  print(4)
elif points['x'] < 0 and points['y'] < 0:
  print(3)
else:
  raise NoQuadrantException

## 2884.	알람 시계

### 문제
    상근이는 매일 아침 알람을 듣고 일어난다. 알람을 듣고 바로 일어나면 다행이겠지만, 항상 조금만 더 자려는 마음 때문에 매일 학교를 지각하고 있다.
    
    상근이는 모든 방법을 동원해보았지만, 조금만 더 자려는 마음은 그 어떤 것도 없앨 수가 없었다.
    
    이런 상근이를 불쌍하게 보던, 창영이는 자신이 사용하는 방법을 추천해 주었다.
    
    바로 "45분 일찍 알람 설정하기"이다.
    
    이 방법은 단순하다. 원래 설정되어 있는 알람을 45분 앞서는 시간으로 바꾸는 것이다. 어차피 알람 소리를 들으면, 알람을 끄고 조금 더 잘 것이기 때문이다. 이 방법을 사용하면, 매일 아침 더 잤다는 기분을 느낄 수 있고, 학교도 지각하지 않게 된다.
    
    현재 상근이가 설정한 알람 시각이 주어졌을 때, 창영이의 방법을 사용한다면, 이를 언제로 고쳐야 하는지 구하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 두 정수 H와 M이 주어진다. (0 ≤ H ≤ 23, 0 ≤ M ≤ 59) 그리고 이것은 현재 상근이가 설정한 놓은 알람 시간 H시 M분을 의미한다.
    
    입력 시간은 24시간 표현을 사용한다. 24시간 표현에서 하루의 시작은 0:0(자정)이고, 끝은 23:59(다음날 자정 1분 전)이다. 시간을 나타낼 때, 불필요한 0은 사용하지 않는다.

### 출력
    첫째 줄에 상근이가 창영이의 방법을 사용할 때, 설정해야 하는 알람 시간을 출력한다. (입력과 같은 형태로 출력하면 된다.)

### 예제 입력 1 
    10 10
### 예제 출력 1 
    9 25
### 예제 입력 2 
    0 30
### 예제 출력 2 
    23 45
### 예제 입력 3 
    23 40
### 예제 출력 3 
    22 55

In [ ]:
H, M = [int(i) for i in input().split()]

if 0 <= H <= 23 and 0 <= M <= 59:
  pass
else:
  raise WrongInputException

if M-45 < 0:
  H = H - 1
  M = 60 - abs(M-45)
else:
  M = M - 45

if H < 0:
  H = 24 + H
else:
  pass

print(H, M)

<a href="https://colab.research.google.com/github/graphy-young/algorithm/blob/main/Baekjoon%20Online%20judge/3.%20for%EB%AC%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 3. for문

for문을 사용해 봅시다.

## 2739. 구구단

### 문제
    N을 입력받은 뒤, 구구단 N단을 출력하는 프로그램을 작성하시오. 출력 형식에 맞춰서 출력하면 된다.

### 입력
    첫째 줄에 N이 주어진다. N은 1보다 크거나 같고, 9보다 작거나 같다.

### 출력
    출력형식과 같게 N*1부터 N*9까지 출력한다.

### 예제 입력 1 
    2
### 예제 출력 1 
    2 * 1 = 2
    2 * 2 = 4
    2 * 3 = 6
    2 * 4 = 8
    2 * 5 = 10
    2 * 6 = 12
    2 * 7 = 14
    2 * 8 = 16
    2 * 9 = 18

### Code

In [ ]:
digit = int(input())

for i in range(1, 10):
  message = "{} * {} = {}".format(digit, i, digit*i)
  print(message)
  #result = digit * i
  #print("{digit} * {i} = {result}")


2
2 * 1 = 2
2 * 2 = 4
2 * 3 = 6
2 * 4 = 8
2 * 5 = 10
2 * 6 = 12
2 * 7 = 14
2 * 8 = 16
2 * 9 = 18


## 10950. A+B - 3

### 문제
    두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 테스트 케이스의 개수 T가 주어진다.
    
    각 테스트 케이스는 한 줄로 이루어져 있으며, 각 줄에 A와 B가 주어진다. (0 < A, B < 10)

### 출력
    각 테스트 케이스마다 A+B를 출력한다.

### 예제 입력 1 
    5
    1 1
    2 3
    3 4
    9 8
    5 2

### 예제 출력 1 
    2
    5
    7
    17
    7

### Code

In [ ]:
testCase = int(input())

for i in range(1, testCase+1):
  digits = list()
  digits = [int(i) for i in input().split()]
  for digit in digits:
    if 0 < digit < 10:
      pass
    else:
      raise WrongInputException
  print(sum(digits))

5
1 1
2
2 3
5
3 4
7
9 8
17
5 2
7


## 8393. 합

### 문제
    n이 주어졌을 때, 1부터 n까지 합을 구하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 n (1 ≤ n ≤ 10,000)이 주어진다.

### 출력
    1부터 n까지 합을 출력한다.

### 예제 입력 1 
    3
### 예제 출력 1 
    6

### Code

In [ ]:
digit = int(input())

if 1 <= digit <= 10000:
  pass
else:
  raise WrongInputException

result = 0

for i in range(1, digit+1):
  result += i

print(result)

## 15552. 빠른 A+B

### 문제
    본격적으로 for문 문제를 풀기 전에 주의해야 할 점이 있다. 입출력 방식이 느리면 여러 줄을 입력받거나 출력할 때 시간초과가 날 수 있다는 점이다.
    
    C++을 사용하고 있고 cin/cout을 사용하고자 한다면, cin.tie(NULL)과 sync_with_stdio(false)를 둘 다 적용해 주고, endl 대신 개행문자(\n)를 쓰자. 단, 이렇게 하면 더 이상 scanf/printf/puts/getchar/putchar 등 C의 입출력 방식을 사용하면 안 된다.
    
    Java를 사용하고 있다면, Scanner와 System.out.println 대신 BufferedReader와 BufferedWriter를 사용할 수 있다. BufferedWriter.flush는 맨 마지막에 한 번만 하면 된다.
    
    Python을 사용하고 있다면, input 대신 sys.stdin.readline을 사용할 수 있다. 단, 이때는 맨 끝의 개행문자까지 같이 입력받기 때문에 문자열을 저장하고 싶을 경우 .rstrip()을 추가로 해 주는 것이 좋다.
    
    또한 입력과 출력 스트림은 별개이므로, 테스트케이스를 전부 입력받아서 저장한 뒤 전부 출력할 필요는 없다. 테스트케이스를 하나 받은 뒤 하나 출력해도 된다.
    
    자세한 설명 및 다른 언어의 경우는 이 글에 설명되어 있다.
    
    이 블로그 글에서 BOJ의 기타 여러 가지 팁을 볼 수 있다.

### 입력
    첫 줄에 테스트케이스의 개수 T가 주어진다. T는 최대 1,000,000이다. 다음 T줄에는 각각 두 정수 A와 B가 주어진다. A와 B는 1 이상, 1,000 이하이다.

### 출력
    각 테스트케이스마다 A+B를 한 줄에 하나씩 순서대로 출력한다.

### 예제 입력 1
    5
    1 1
    12 34
    5 500
    40 60
    1000 1000

### 예제 출력 1
    2
    46
    505
    100
    2000

### Code

In [ ]:
# this doesn't work in colab
from sys import stdin

t = int(stdin.readline())

for i in range(t):
  digits = map(int, stdin.readline().split())
  result = sum(digits)
  print(result)

## 2741. N 찍기

### 문제
    자연수 N이 주어졌을 때, 1부터 N까지 한 줄에 하나씩 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 100,000보다 작거나 같은 자연수 N이 주어진다.

### 출력
    첫째 줄부터 N번째 줄 까지 차례대로 출력한다.

### 예제 입력 1
    5

### 예제 출력 1
    1
    2
    3
    4
    5

### Code

In [ ]:
from sys import stdin
N = int(stdin.readline().rstrip())

for i in range(N):
  print(i+1)

## 2742. 기찍 N

### 문제
    자연수 N이 주어졌을 때, N부터 1까지 한 줄에 하나씩 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 100,000보다 작거나 같은 자연수 N이 주어진다.

### 출력
    첫째 줄부터 N번째 줄 까지 차례대로 출력한다.

### 예제 입력 1
    5
### 예제 출력 1
    5
    4
    3
    2
    1

### Code

In [ ]:
# N = int(input())
from sys import stdin
N = int(stdin.readline())

for i in range(N, 0, -1):
  print(i)

## 11021. A+B - 7

### 문제
    두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 테스트 케이스의 개수 T가 주어진다.
    
    각 테스트 케이스는 한 줄로 이루어져 있으며, 각 줄에 A와 B가 주어진다. (0 < A, B < 10)

### 출력
    각 테스트 케이스마다 "Case #x: "를 출력한 다음, A+B를 출력한다. 테스트 케이스 번호는 1부터 시작한다.

### 예제 입력 1
    5
    1 1
    2 3
    3 4
    9 8
    5 2

### 예제 출력 1
    Case #1: 2
    Case #2: 5
    Case #3: 7
    Case #4: 17
    Case #5: 7

### Code


In [ ]:
from sys import stdin
t = int(stdin.readline())

for i in range(t):
  digits = map(int, stdin.readline().split())
  print("Case #{}:".format(i+1), str(sum(digits)))

## 11022. A+B - 8

### 문제
    두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 테스트 케이스의 개수 T가 주어진다.
    
    각 테스트 케이스는 한 줄로 이루어져 있으며, 각 줄에 A와 B가 주어진다. (0 < A, B < 10)

### 출력
    각 테스트 케이스마다 "Case #x: A + B = C" 형식으로 출력한다. x는 테스트 케이스 번호이고 1부터 시작하며, C는 A+B이다.

### 예제 입력 1
    5
    1 1
    2 3
    3 4
    9 8
    5 2

### 예제 출력 1
    Case #1: 1 + 1 = 2
    Case #2: 2 + 3 = 5
    Case #3: 3 + 4 = 7
    Case #4: 9 + 8 = 17
    Case #5: 5 + 2 = 7

### Code

In [ ]:
from sys import stdin
t = int(stdin.readline())

for i in range(t):
  digits = list(map(int, stdin.readline().split()))
  print("Case #{}: {} + {} = {}".format(i+1, digits[0], digits[1], sum(digits)))

## 2438	별 찍기 - 1

### 문제
    첫째 줄에는 별 1개, 둘째 줄에는 별 2개, N번째 줄에는 별 N개를 찍는 문제

### 입력
    첫째 줄에 N(1 ≤ N ≤ 100)이 주어진다.

### 출력
    첫째 줄부터 N번째 줄까지 차례대로 별을 출력한다.

### 예제 입력 1
    5

### 예제 출력 1
    *
    **
    ***
    ****
    *****

### Code

In [ ]:
from sys import stdin, stdout
n = int(stdin.readline())

for i in range(n):
  for j in range(i+1):
    stdout.write('*')
  if i+1 != n: 
    stdout.write('\n')

## 2439. 별 찍기 - 2

### 문제
    첫째 줄에는 별 1개, 둘째 줄에는 별 2개, N번째 줄에는 별 N개를 찍는 문제
    
    하지만, 오른쪽을 기준으로 정렬한 별(예제 참고)을 출력하시오.

### 입력
    첫째 줄에 N(1 ≤ N ≤ 100)이 주어진다.

### 출력
    첫째 줄부터 N번째 줄까지 차례대로 별을 출력한다.

### 예제 입력 1
    5
### 예제 출력 1
        *
       **
      ***
     ****
    *****

### Code

In [ ]:
from sys import stdin
n = int(stdin.readline())
#n = int(input())

for i in range(n):
  result = ""
  for j in range(i+1):
    result += "*"
  print(result.rjust(n))

## 10871. X보다 작은 수

### 문제
    정수 N개로 이루어진 수열 A와 정수 X가 주어진다. 이때, A에서 X보다 작은 수를 모두 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 N과 X가 주어진다. (1 ≤ N, X ≤ 10,000)
    
    둘째 줄에 수열 A를 이루는 정수 N개가 주어진다. 주어지는 정수는 모두 1보다 크거나 같고, 10,000보다 작거나 같은 정수이다.

### 출력
    X보다 작은 수를 입력받은 순서대로 공백으로 구분해 출력한다. X보다 작은 수는 적어도 하나 존재한다.

### 예제 입력 1
    10 5
    1 10 4 9 2 3 8 5 7 6

### 예제 출력 1
    1 4 2 3

### Code

In [ ]:
from sys import stdin
n, x = list(map(int, stdin.readline().split()))

a = list(map(int, stdin.readline().split()))
result = list()

for i in a:
  if i < x:
    result.append(str(i))
  else:
    pass

print(' '.join(result))

<a href="https://colab.research.google.com/github/graphy-young/algorithm/blob/main/Baekjoon%20Online%20Judge/4_While%EB%AC%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 4. While문

while문을 사용해 봅시다.

## 10952. A+B - 5

### 문제
    두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

### 입력
    입력은 여러 개의 테스트 케이스로 이루어져 있다.
    
    각 테스트 케이스는 한 줄로 이루어져 있으며, 각 줄에 A와 B가 주어진다. (0 < A, B < 10)
    
    입력의 마지막에는 0 두 개가 들어온다.

### 출력
    각 테스트 케이스마다 A+B를 출력한다.

### 예제 입력 1
    1 1
    2 3
    3 4
    9 8
    5 2
    0 0

### 예제 출력 1
    2
    5
    7
    17
    7

### Code

In [ ]:
from sys import stdin

while(True):
  #A, B = list(map(int, stdin.readline().split()))
  A, B = list(map(int, input().split()))
  if A == 0 and B == 0:
    break
  else:
    print(A+B)

1 1
2
2 3
5
3 4
7
9 8
17
5 2
7
0 0


## 10951. A+B - 4

### 문제
    두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

### 입력
    입력은 여러 개의 테스트 케이스로 이루어져 있다.
    
    각 테스트 케이스는 한 줄로 이루어져 있으며, 각 줄에 A와 B가 주어진다. (0 < A, B < 10)

### 출력
    각 테스트 케이스마다 A+B를 출력한다.

### 예제 입력 1
    1 1
    2 3
    3 4
    9 8
    5 2

### 예제 출력 1
    2
    5
    7
    17
    7

### Code

EOF에 대응하는 방법을 알아야는데, 마땅히 괜찮은 방법이 인터넷에 안보임..

In [ ]:
from sys import stdin

while(True):
  try:
    A, B = list(map(int, stdin.readline().split()))
    #A, B = list(map(int, input().split()))
  except ValueError:
    break
  print(A+B)

## 1110. 더하기 사이클

### 입력
    첫째 줄에 N이 주어진다. N은 0보다 크거나 같고, 99보다 작거나 같은 정수이다.

### 출력
    첫째 줄에 N의 사이클 길이를 출력한다.

### 예제 입력 1 
    26
### 예제 출력 1 
    4
### 예제 입력 2 
    55
### 예제 출력 2 
     3
### 예제 입력 3 
    1
### 예제 출력 3 
    60
### 예제 입력 4 
    0
### 예제 출력 4
    1

### Code

In [ ]:
digit = int(input())

if digit < 10:
  digit = digit * 10

tens = digit // 10
units = digit % 10
cnt = 0

while(True):
  tens, units = units, tens
  units = tens + units
  if units >= 10: 
    units = units % 10
  cnt += 1
  if digit == tens*10 + units:
    break

print(cnt)

26
4


<a href="https://colab.research.google.com/github/graphy-young/algorithm/blob/main/Baekjoon%20Online%20Judge/5_1%EC%B0%A8%EC%9B%90%20%EB%B0%B0%EC%97%B4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 5. 1차원 배열

배열을 사용해 봅시다.

## 10818. 최소, 최대

### 문제
    N개의 정수가 주어진다. 이때, 최솟값과 최댓값을 구하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 정수의 개수 N (1 ≤ N ≤ 1,000,000)이 주어진다. 둘째 줄에는 N개의 정수를 공백으로 구분해서 주어진다. 모든 정수는 -1,000,000보다 크거나 같고, 1,000,000보다 작거나 같은 정수이다.

### 출력
    첫째 줄에 주어진 정수 N개의 최솟값과 최댓값을 공백으로 구분해 출력한다.

### 예제 입력 1 
    5
    20 10 35 30 7
### 예제 출력 1 
    7 35

### Code

In [ ]:
digitNums = int(input())
numbers = list(map(int, input().split()))

print(min(numbers), max(numbers))

5
20 10 35 30 7
35 7


## 2562. 최댓값

### 문제
    9개의 서로 다른 자연수가 주어질 때, 이들 중 최댓값을 찾고 그 최댓값이 몇 번째 수인지를 구하는 프로그램을 작성하시오.
    
    예를 들어, 서로 다른 9개의 자연수
    
    3, 29, 38, 12, 57, 74, 40, 85, 61
    
    이 주어지면, 이들 중 최댓값은 85이고, 이 값은 8번째 수이다.

### 입력
    첫째 줄부터 아홉 번째 줄까지 한 줄에 하나의 자연수가 주어진다. 주어지는 자연수는 100 보다 작다.

### 출력
    첫째 줄에 최댓값을 출력하고, 둘째 줄에 최댓값이 몇 번째 수인지를 출력한다.

### 예제 입력 1 
    3
    29
    38
    12
    57
    74
    40
    85
    61

### 예제 출력 1 
    85
    8

### Code

In [ ]:
maxValue = 0
cnt = 0

for i in range(9):
  digit = int(input())
  if maxValue < digit:
    maxValue = digit
    cnt = i + 1  
  else:
    continue
else:
  print(maxValue)
  print(cnt)

3
29
38
12
57
74
40
85
61
85
8


## 2577. 숫자의 개수

### 문제
    세 개의 자연수 A, B, C가 주어질 때 A×B×C를 계산한 결과에 0부터 9까지 각각의 숫자가 몇 번씩 쓰였는지를 구하는 프로그램을 작성하시오.
    
    예를 들어 A = 150, B = 266, C = 427 이라면 
    
    A × B × C = 150 × 266 × 427 = 17037300 이 되고, 
    
    계산한 결과 17037300 에는 0이 3번, 1이 1번, 3이 2번, 7이 2번 쓰였다.

### 입력
    첫째 줄에 A, 둘째 줄에 B, 셋째 줄에 C가 주어진다. A, B, C는 모두 100보다 같거나 크고, 1,000보다 작은 자연수이다.

### 출력
    첫째 줄에는 A×B×C의 결과에 0 이 몇 번 쓰였는지 출력한다. 마찬가지로 둘째 줄부터 열 번째 줄까지 A×B×C의 결과에 1부터 9까지의 숫자가 각각 몇 번 쓰였는지 차례로 한 줄에 하나씩 출력한다.

### 예제 입력 1 
    150
    266
    427


### 예제 출력 1 
    3
    1
    0
    2
    0
    0
    0
    2
    0
    0

### Code

In [ ]:
numbers = 1
for i in range(3):
  numbers *= int(input())
numbers = list(map(int, list(str(numbers))))

for digit in range(10):
  cnt = 0
  for number in numbers:
    if number == digit:
      cnt += 1
  print(cnt)

150
266
427
[1, 7, 0, 3, 7, 3, 0, 0]
3
1
0
2
0
0
0
2
0
0


## 3052. 나머지

### 문제
    두 자연수 A와 B가 있을 때, A%B는 A를 B로 나눈 나머지 이다. 예를 들어, 7, 14, 27, 38을 3으로 나눈 나머지는 1, 2, 0, 2이다. 
    
    수 10개를 입력받은 뒤, 이를 42로 나눈 나머지를 구한다. 그 다음 서로 다른 값이 몇 개 있는지 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄부터 열번째 줄 까지 숫자가 한 줄에 하나씩 주어진다. 이 숫자는 1,000보다 작거나 같고, 음이 아닌 정수이다.

### 출력
    첫째 줄에, 42로 나누었을 때, 서로 다른 나머지가 몇 개 있는지 출력한다.

### 예제 입력 1 
    1
    2
    3
    4
    5
    6
    7
    8
    9
    10

### 예제 출력 1 
    10
각 수를 42로 나눈 나머지는 1, 2, 3, 4, 5, 6, 7, 8, 9, 10이다.

### 예제 입력 2 
    42
    84
    252
    420
    840
    126
    42
    84
    420
    126

### 예제 출력 2 
    1
모든 수를 42로 나눈 나머지는 0이다.

### 예제 입력 3 
    39
    40
    41
    42
    43
    44
    82
    83
    84
    85

### 예제 출력 3 
    6
각 수를 42로 나눈 나머지는 39, 40, 41, 0, 1, 2, 40, 41, 0, 1이다. 서로 다른 값은 6개가 있다.

### Code

In [ ]:
numbers = list()

for i in range(10):
  numbers.append(int(input()))
  numbers[i] = numbers[i] % 42

numbers = list(set(numbers))
print(len(numbers))

39
40
41
42
43
44
82
83
84
85
6


## 1546. 평균

### 문제
    세준이는 기말고사를 망쳤다. 세준이는 점수를 조작해서 집에 가져가기로 했다. 일단 세준이는 자기 점수 중에 최댓값을 골랐다. 이 값을 M이라고 한다. 그리고 나서 모든 점수를 점수/M*100으로 고쳤다.
    
    예를 들어, 세준이의 최고점이 70이고, 수학점수가 50이었으면 수학점수는 50/70*100이 되어 71.43점이 된다.
    
    세준이의 성적을 위의 방법대로 새로 계산했을 때, 새로운 평균을 구하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 시험 본 과목의 개수 N이 주어진다. 이 값은 1000보다 작거나 같다. 둘째 줄에 세준이의 현재 성적이 주어진다. 이 값은 100보다 작거나 같은 음이 아닌 정수이고, 적어도 하나의 값은 0보다 크다.

### 출력
    첫째 줄에 새로운 평균을 출력한다. 실제 정답과 출력값의 절대오차 또는 상대오차가 10-2 이하이면 정답이다.

### 예제 입력 1 
    3
    40 80 60

### 예제 출력 1 
    75.0

### 예제 입력 2 
    3
    10 20 30

### 예제 출력 2 
    66.666667
10-2 이하의 오차를 허용한다는 말은 정확히 소수 2번째 자리까지 출력하라는 뜻이 아니다.

### 예제 입력 3 
    4
    1 100 100 100

### 예제 출력 3 
    75.25

### 예제 입력 4 
    5
    1 2 4 8 16

### 예제 출력 4 
    38.75
### 예제 입력 5 
    2
    3 10

### 예제 출력 5 
    65.0

### Code

In [ ]:
subjectNumber = int(input())
scores = list(map(int, input().split()))

maxScore = max(scores)
for score in scores:
  scores[scores.index(score)] = score / maxScore * 100

avg = sum(scores) / len(scores)
print(avg)

3
40 80 60
[50.0, 100.0, 75.0]
75.0


## 8958. OX퀴즈

### 문제
    "OOXXOXXOOO"와 같은 OX퀴즈의 결과가 있다. O는 문제를 맞은 것이고, X는 문제를 틀린 것이다. 문제를 맞은 경우 그 문제의 점수는 그 문제까지 연속된 O의 개수가 된다. 예를 들어, 10번 문제의 점수는 3이 된다.
    
    "OOXXOXXOOO"의 점수는 1+2+0+0+1+0+0+1+2+3 = 10점이다.
    
    OX퀴즈의 결과가 주어졌을 때, 점수를 구하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 테스트 케이스의 개수가 주어진다. 각 테스트 케이스는 한 줄로 이루어져 있고, 길이가 0보다 크고 80보다 작은 문자열이 주어진다. 문자열은 O와 X만으로 이루어져 있다.

### 출력
    각 테스트 케이스마다 점수를 출력한다.

### 예제 입력 1 
    5
    OOXXOXXOOO
    OOXXOOXXOO
    OXOXOXOXOXOXOX
    OOOOOOOOOO
    OOOOXOOOOXOOOOX

### 예제 출력 1 
    10
    9
    7
    55
    30

### Code

In [ ]:
testCase = int(input())

for i in range(testCase):
  scores = list(input())
  cumCnt = 0
  result = 0
  for score in scores:
    if score == 'O':
      cumCnt += 1
    else:
      cumCnt = 0
    result += cumCnt
  print(result)

5
OOXXOXXOOO
10
OOXXOOXXOO
9
OXOXOXOXOXOXOX
7
OOOOOOOOOO
55
OOOOXOOOOXOOOOX
30


## 4344. 평균은 넘겠지

### 문제
    대학생 새내기들의 90%는 자신이 반에서 평균은 넘는다고 생각한다. 당신은 그들에게 슬픈 진실을 알려줘야 한다.

### 입력
    첫째 줄에는 테스트 케이스의 개수 C가 주어진다.
    
    둘째 줄부터 각 테스트 케이스마다 학생의 수 N(1 ≤ N ≤ 1000, N은 정수)이 첫 수로 주어지고, 이어서 N명의 점수가 주어진다. 점수는 0보다 크거나 같고, 100보다 작거나 같은 정수이다.

### 출력
    각 케이스마다 한 줄씩 평균을 넘는 학생들의 비율을 반올림하여 소수점 셋째 자리까지 출력한다.

### 예제 입력 1 
    5
    5 50 50 70 80 100
    7 100 95 90 80 70 60 50
    3 70 90 80
    3 70 90 81
    9 100 99 98 97 96 95 94 93 91

### 예제 출력 1 
    40.000%
    57.143%
    33.333%
    66.667%
    55.556%

### Code

In [ ]:
#from sys import stdin

testCase = int(input())
#testCase = int(stdin.readline().rstrip())

for i in range(testCase):
  scores = list(map(int, input().split(' ')))
  #scores = list(map(int, stdin.readline().split()))
  groupAvg = sum(scores[1:]) / scores[0]
  for score in scores[1:]:
    if score > groupAvg:
      scores[scores.index(score)] = True
    else:
      scores[scores.index(score)] = False
  aboveRatio = round(sum(scores[1:]) / scores[0] * 100, 3)
  print("%0.3f" % aboveRatio + "%")

5
5 50 50 70 80 100
40.000%
7 100 95 90 80 70 60 50
57.143%
3 70 90 80
33.333%
3 70 90 81
66.667%
9 100 99 98 97 96 95 94 93 91
55.556%


** Colab에서는 정상 작동하나, 백준 제출에서는 런타임 에러 발생. 수정 필요할듯 **

<a href="https://colab.research.google.com/github/graphy-young/algorithm/blob/main/Baekjoon%20Online%20Judge/6_%ED%95%A8%EC%88%98.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 6. 함수

함수를 정의하면 코드가 깔끔해지고 관리하기 쉬워집니다.

## 15596. 정수 N개의 합

### 문제
    정수 n개가 주어졌을 때, n개의 합을 구하는 함수를 작성하시오.
    
    작성해야 하는 함수는 다음과 같다.
    
    C, C11, C (Clang), C11 (Clang): long long sum(int *a, int n);
    a: 합을 구해야 하는 정수 n개가 저장되어 있는 배열 (0 ≤ a[i] ≤ 1,000,000, 1 ≤ n ≤ 3,000,000)
    n: 합을 구해야 하는 정수의 개수
    리턴값: a에 포함되어 있는 정수 n개의 합
    
    C++, C++11, C++14, C++17, C++ (Clang), C++11 (Clang), C++14 (Clang), C++17 (Clang): long long sum(std::vector<int> &a);
    a: 합을 구해야 하는 정수 n개가 저장되어 있는 배열 (0 ≤ a[i] ≤ 1,000,000, 1 ≤ n ≤ 3,000,000)
    리턴값: a에 포함되어 있는 정수 n개의 합
    
    Python 2, Python 3, PyPy, PyPy3: def solve(a: list) -> int
    a: 합을 구해야 하는 정수 n개가 저장되어 있는 리스트 (0 ≤ a[i] ≤ 1,000,000, 1 ≤ n ≤ 3,000,000)
    리턴값: a에 포함되어 있는 정수 n개의 합 (정수)
    
    Java: long sum(int[] a); (클래스 이름: Test)
    a: 합을 구해야 하는 정수 n개가 저장되어 있는 배열 (0 ≤ a[i] ≤ 1,000,000, 1 ≤ n ≤ 3,000,000)
    리턴값: a에 포함되어 있는 정수 n개의 합
    
    Go: sum(a []int) int
    a: 합을 구해야 하는 정수 n개가 저장되어 있는 배열 (0 ≤ a[i] ≤ 1,000,000, 1 ≤ n ≤ 3,000,000)
    리턴값: a에 포함되어 있는 정수 n개의 합

### Code

In [ ]:
def solve(a):
    return sum(a)

## 4673. 셀프 넘버

### 문제
    셀프 넘버는 1949년 인도 수학자 D.R. Kaprekar가 이름 붙였다. 양의 정수 n에 대해서 d(n)을 n과 n의 각 자리수를 더하는 함수라고 정의하자. 예를 들어, d(75) = 75+7+5 = 87이다.
    
    양의 정수 n이 주어졌을 때, 이 수를 시작해서 n, d(n), d(d(n)), d(d(d(n))), ...과 같은 무한 수열을 만들 수 있다. 
    
    예를 들어, 33으로 시작한다면 다음 수는 33 + 3 + 3 = 39이고, 그 다음 수는 39 + 3 + 9 = 51, 다음 수는 51 + 5 + 1 = 57이다. 이런식으로 다음과 같은 수열을 만들 수 있다.
    
    33, 39, 51, 57, 69, 84, 96, 111, 114, 120, 123, 129, 141, ...
    
    n을 d(n)의 생성자라고 한다. 위의 수열에서 33은 39의 생성자이고, 39는 51의 생성자, 51은 57의 생성자이다. 생성자가 한 개보다 많은 경우도 있다. 예를 들어, 101은 생성자가 2개(91과 100) 있다. 
    
    생성자가 없는 숫자를 셀프 넘버라고 한다. 100보다 작은 셀프 넘버는 총 13개가 있다. 1, 3, 5, 7, 9, 20, 31, 42, 53, 64, 75, 86, 97
    
    10000보다 작거나 같은 셀프 넘버를 한 줄에 하나씩 출력하는 프로그램을 작성하시오.

### 입력
    입력은 없다.

### 출력
    10,000보다 작거나 같은 셀프 넘버를 한 줄에 하나씩 증가하는 순서로 출력한다.

### 예제 출력 1 
    1
    3
    5
    7
    9
    20
    31
    42
    53
    64
     |
     |       <-- a lot more numbers
     |
    9903
    9914
    9925
    9927
    9938
    9949
    9960
    9971
    9982
    9993

### Code

#### 1st try. 시간초과

오답인거같음...

In [ ]:
def getSelfNumber(num, numList=[]):
  selfNumber = num + (num // 10) + (num % 10)
  if selfNumber <= 10000:
    numList.append(selfNumber)
    return getSelfNumber(selfNumber, numList)
  else:
    return numList

targetList = [i for i in range(1, 10001)]
selfNumList = set()

for i in range(1, 10001):
  selfNumList.update(getSelfNumber(i))

result = sorted(list(set(targetList) - selfNumList))

for j in result:
  print(j)

1
3
5
7
9
20
31
42
53
64
75
86
97
108
119
130
141
152
163
174
185
196
207
218
229
240
251
262
273
284
295
306
317
328
339
350
361
372
383
394
405
416
427
438
449
460
471
482
493
504
515
526
537
548
559
570
581
592
603
614
625
636
647
658
669
680
691
702
713
724
735
746
757
768
779
790
801
812
823
834
845
856
867
878
889
900
911
922
933
944
955
966
977
988
999
1010
1021
1032
1043
1054
1065
1076
1087
1098
1109
1120
1131
1142
1153
1164
1175
1186
1197
1208
1219
1230
1241
1252
1263
1274
1285
1296
1307
1318
1329
1340
1351
1362
1373
1384
1395
1406
1417
1428
1439
1450
1461
1472
1483
1494
1505
1516
1527
1538
1549
1560
1571
1582
1593
1604
1615
1626
1637
1648
1659
1670
1681
1692
1703
1714
1725
1736
1747
1758
1769
1780
1791
1802
1813
1824
1835
1846
1857
1868
1879
1890
1901
1912
1923
1934
1945
1956
1967
1978
1989
2000
2011
2022
2033
2044
2055
2066
2077
2088
2099
2110
2121
2132
2143
2154
2165
2176
2187
2198
2209
2220
2231
2242
2253
2264
2275
2286
2297
2308
2319
2330
2341
2352
2363
2374
2385
2396
240

#### 2nd try

In [ ]:
def showSelfNumbers(selfNumList=[]):
  
  maxValue = 10000

  if selfNumList == []:
    selfNumList = [i for i in range(1, maxValue+1)]

  def generateNumber(number):
    digits = [ int(i) for i in list(str(number)) ]
    return number + sum(digits)
  
  while(selfNumList != []):
    number = selfNumList[0]
    print(number)
    selfNumList.remove(number)

    generatedNumber = generateNumber(number)
    while(generatedNumber <= maxValue):
      if generatedNumber in selfNumList:
        selfNumList.remove(generatedNumber)
      else:
        break
      generatedNumber = generateNumber(generatedNumber)

showSelfNumbers()

1
3
5
7
9
20
31
42
53
64
75
86
97
108
110
121
132
143
154
165
176
187
198
209
211
222
233
244
255
266
277
288
299
310
312
323
334
345
356
367
378
389
400
411
413
424
435
446
457
468
479
490
501
512
514
525
536
547
558
569
580
591
602
613
615
626
637
648
659
670
681
692
703
714
716
727
738
749
760
771
782
793
804
815
817
828
839
850
861
872
883
894
905
916
918
929
940
951
962
973
984
995
1006
1021
1032
1043
1054
1065
1076
1087
1098
1109
1111
1122
1133
1144
1155
1166
1177
1188
1199
1210
1212
1223
1234
1245
1256
1267
1278
1289
1300
1311
1313
1324
1335
1346
1357
1368
1379
1390
1401
1412
1414
1425
1436
1447
1458
1469
1480
1491
1502
1513
1515
1526
1537
1548
1559
1570
1581
1592
1603
1614
1616
1627
1638
1649
1660
1671
1682
1693
1704
1715
1717
1728
1739
1750
1761
1772
1783
1794
1805
1816
1818
1829
1840
1851
1862
1873
1884
1895
1906
1917
1919
1930
1941
1952
1963
1974
1985
1996
2007
2022
2033
2044
2055
2066
2077
2088
2099
2110
2112
2123
2134
2145
2156
2167
2178
2189
2200
2211
2213
2224
2235
2246


## 1065. 한수

### 문제
    어떤 양의 정수 X의 각 자리가 등차수열을 이룬다면, 그 수를 한수라고 한다. 등차수열은 연속된 두 개의 수의 차이가 일정한 수열을 말한다. N이 주어졌을 때, 1보다 크거나 같고, N보다 작거나 같은 한수의 개수를 출력하는 프로그램을 작성하시오. 

### 입력
    첫째 줄에 1,000보다 작거나 같은 자연수 N이 주어진다.

### 출력
    첫째 줄에 1보다 크거나 같고, N보다 작거나 같은 한수의 개수를 출력한다.

### 예제 입력 1 
    110
### 예제 출력 1 
    99
### 예제 입력 2 
    1
### 예제 출력 2 
    1
### 예제 입력 3 
    210
### 예제 출력 3 
    105
### 예제 입력 4 
    1000
### 예제 출력 4 
    144

### Code

In [ ]:
# I don't know what means hansu in english..
def getAllHansu(maxRange):
  cnt = 0
  for i in range(1, maxRange+1):
    digits = [ int(d) for d in list(str(i)) ]
    if len(digits) <= 2:
      cnt += 1
    else:
      apList = []
      for j in range(len(digits)-1):
        apList.append(digits[j] - digits[j+1])
      for k in range(len(apList)-1):
        if apList[k] == apList[k+1]:
          continue
        else:
          break
      else:
        cnt += 1
  print(cnt)


getAllHansu(int(input()))

1000
144


# 7. 문자열

문자열을 다루는 문제들을 해결해 봅시다.

## 11654. 아스키 코드

### 문제
    알파벳 소문자, 대문자, 숫자 0-9중 하나가 주어졌을 때, 주어진 글자의 아스키 코드값을 출력하는 프로그램을 작성하시오.

### 입력
    알파벳 소문자, 대문자, 숫자 0-9 중 하나가 첫째 줄에 주어진다.

### 출력
    입력으로 주어진 글자의 아스키 코드 값을 출력한다.

### 예제 입력 1 
    A
### 예제 출력 1 
    65
### 예제 입력 2 
    C
### 예제 출력 2 
    67
### 예제 입력 3 
    0
### 예제 출력 3 
    48
### 예제 입력 4 
    9
### 예제 출력 4 
    57
### 예제 입력 5 
    a
### 예제 출력 5 
    97
### 예제 입력 6 
    z
### 예제 출력 6 
    122

### Code

In [ ]:
userInput = ord(input())
print(userInput)

A
65


## 11720. 숫자의 합

### 문제
    N개의 숫자가 공백 없이 쓰여있다. 이 숫자를 모두 합해서 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 숫자의 개수 N (1 ≤ N ≤ 100)이 주어진다. 둘째 줄에 숫자 N개가 공백없이 주어진다.

### 출력
    입력으로 주어진 숫자 N개의 합을 출력한다.

### 예제 입력 1 
    1
    1
### 예제 출력 1 
    1
### 예제 입력 2 
    5
    54321
### 예제 출력 2 
    15
### 예제 입력 3 
    25
    7000000000000000000000000
### 예제 출력 3 
    7
### 예제 입력 4 
    11
    10987654321
### 예제 출력 4 
    46

### Code

In [ ]:
numDigit = int(input())
digitList = [int(i) for i in list(input())]

print(sum(digitList))

## 10809. 알파벳 찾기

### 문제
    알파벳 소문자로만 이루어진 단어 S가 주어진다. 각각의 알파벳에 대해서, 단어에 포함되어 있는 경우에는 처음 등장하는 위치를, 포함되어 있지 않은 경우에는 -1을 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 단어 S가 주어진다. 단어의 길이는 100을 넘지 않으며, 알파벳 소문자로만 이루어져 있다.

### 출력
    각각의 알파벳에 대해서, a가 처음 등장하는 위치, b가 처음 등장하는 위치, ... z가 처음 등장하는 위치를 공백으로 구분해서 출력한다.
    
    만약, 어떤 알파벳이 단어에 포함되어 있지 않다면 -1을 출력한다. 단어의 첫 번째 글자는 0번째 위치이고, 두 번째 글자는 1번째 위치이다.

### 예제 입력 1 
    baekjoon
### 예제 출력 1 
    1 0 -1 -1 2 -1 -1 -1 -1 4 3 -1 -1 7 5 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1

### Code

In [ ]:
S = input()
alphabetList = { i:-1 for i in list('abcdefghijklmnopqrstuvwxyz') }

for alphabet in S:
  if alphabetList[alphabet] == -1:
    alphabetList[alphabet] = S.index(alphabet)

result = [str(i) for i in alphabetList.values()]
print(' '.join(result))

## 2675. 문자열 반복

### 문제
    문자열 S를 입력받은 후에, 각 문자를 R번 반복해 새 문자열 P를 만든 후 출력하는 프로그램을 작성하시오. 즉, 첫 번째 문자를 R번 반복하고, 두 번째 문자를 R번 반복하는 식으로 P를 만들면 된다. S에는 QR Code "alphanumeric" 문자만 들어있다.
    
    QR Code "alphanumeric" 문자는 0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ\$%*+-./: 이다.

### 입력
    첫째 줄에 테스트 케이스의 개수 T(1 ≤ T ≤ 1,000)가 주어진다. 각 테스트 케이스는 반복 횟수 R(1 ≤ R ≤ 8), 문자열 S가 공백으로 구분되어 주어진다. S의 길이는 적어도 1이며, 20글자를 넘지 않는다. 

### 출력
    각 테스트 케이스에 대해 P를 출력한다.

### 예제 입력 1 
    2
    3 ABC
    5 /HTP
### 예제 출력 1 
    AAABBBCCC
    /////HHHHHTTTTTPPPPP

### Code

In [ ]:
T = int(input())

for i in range(T):
  S = input()
  repeatNum, word = S.split()
  resultList = str()

  for cha in list(word):
    resultList += cha * int(repeatNum)

  print(''.join(resultList))

## 1157. 단어 공부

### 문제
    알파벳 대소문자로 된 단어가 주어지면, 이 단어에서 가장 많이 사용된 알파벳이 무엇인지 알아내는 프로그램을 작성하시오. 단, 대문자와 소문자를 구분하지 않는다.

### 입력
    첫째 줄에 알파벳 대소문자로 이루어진 단어가 주어진다. 주어지는 단어의 길이는 1,000,000을 넘지 않는다.

### 출력
    첫째 줄에 이 단어에서 가장 많이 사용된 알파벳을 대문자로 출력한다. 단, 가장 많이 사용된 알파벳이 여러 개 존재하는 경우에는 ?를 출력한다.

### 예제 입력 1 
    Mississipi
### 예제 출력 1 
    ?
### 예제 입력 2 
    zZa
### 예제 출력 2 
    Z
### 예제 입력 3 
    z
### 예제 출력 3 
    Z
### 예제 입력 4 
    baaa
### 예제 출력 4 
    A

### Code

In [ ]:
word = input().upper()
alphabetDict = { i:0 for i in 'ABCDEFGHIJKLMNOPQRSTUVWXYZ' }

for character in word:
  alphabetDict[character] += 1

maxValue = max(alphabetDict.values())

if list(alphabetDict.values()).count(maxValue) > 1:
  print('?')
else:
  print(str([k for k, v in alphabetDict.items() if v == maxValue][0]))

## 1152. 단어의 개수

### 문제
    영어 대소문자와 띄어쓰기만으로 이루어진 문자열이 주어진다. 이 문자열에는 몇 개의 단어가 있을까? 이를 구하는 프로그램을 작성하시오. 단, 한 단어가 여러 번 등장하면 등장한 횟수만큼 모두 세어야 한다.

### 입력
    첫 줄에 영어 대소문자와 띄어쓰기로 이루어진 문자열이 주어진다. 이 문자열의 길이는 1,000,000을 넘지 않는다. 단어는 띄어쓰기 한 개로 구분되며, 공백이 연속해서 나오는 경우는 없다. 또한 문자열의 앞과 뒤에는 공백이 있을 수도 있다.

### 출력
    첫째 줄에 단어의 개수를 출력한다.

### 예제 입력 1 
    The Curious Case of Benjamin Button
### 예제 출력 1 
    6
### 예제 입력 2 
    Mazatneunde Wae Teullyeoyo
### 예제 출력 2 
    3
### 예제 입력 3 
    Teullinika Teullyeotzi 
### 예제 출력 3 
    2

### Code

In [ ]:
userInput = input()
listLength = len(list(filter(('').__ne__, userInput.split(' '))))
print(listLength)

 where where
2


## 2908. 상수

### 문제
    상근이의 동생 상수는 수학을 정말 못한다. 상수는 숫자를 읽는데 문제가 있다. 이렇게 수학을 못하는 상수를 위해서 상근이는 수의 크기를 비교하는 문제를 내주었다. 상근이는 세 자리 수 두 개를 칠판에 써주었다. 그 다음에 크기가 큰 수를 말해보라고 했다.

    상수는 수를 다른 사람과 다르게 거꾸로 읽는다. 예를 들어, 734와 893을 칠판에 적었다면, 상수는 이 수를 437과 398로 읽는다. 따라서, 상수는 두 수중 큰 수인 437을 큰 수라고 말할 것이다.

    두 수가 주어졌을 때, 상수의 대답을 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 상근이가 칠판에 적은 두 수 A와 B가 주어진다. 두 수는 같지 않은 세 자리 수이며, 0이 포함되어 있지 않다.

### 출력
    첫째 줄에 상수의 대답을 출력한다.

### 예제 입력 1 
    734 893
### 예제 출력 1 
    437

### Code

In [ ]:
userInput = input().split(' ')

for num in userInput:
  userInput[userInput.index(num)] = int(num[::-1])

print(max(userInput))

123 321
321


## 5622. 다이얼

#### 그림

https://onlinejudgeimages.s3-ap-northeast-1.amazonaws.com/upload/images/dial.png

### 문제
    상근이의 할머니는 아래 그림과 같이 오래된 다이얼 전화기를 사용한다.
    
    전화를 걸고 싶은 번호가 있다면, 숫자를 하나를 누른 다음에 금속 핀이 있는 곳 까지 시계방향으로 돌려야 한다. 숫자를 하나 누르면 다이얼이 처음 위치로 돌아가고, 다음 숫자를 누르려면 다이얼을 처음 위치에서 다시 돌려야 한다.
    
    숫자 1을 걸려면 총 2초가 필요하다. 1보다 큰 수를 거는데 걸리는 시간은 이보다 더 걸리며, 한 칸 옆에 있는 숫자를 걸기 위해선 1초씩 더 걸린다.
    
    상근이의 할머니는 전화 번호를 각 숫자에 해당하는 문자로 외운다. 즉, 어떤 단어를 걸 때, 각 알파벳에 해당하는 숫자를 걸면 된다. 예를 들어, UNUCIC는 868242와 같다.
    
    할머니가 외운 단어가 주어졌을 때, 이 전화를 걸기 위해서 필요한 최소 시간을 구하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 알파벳 대문자로 이루어진 단어가 주어진다. 단어의 길이는 2보다 크거나 같고, 15보다 작거나 같다.

### 출력
    첫째 줄에 다이얼을 걸기 위해서 필요한 최소 시간을 출력한다.

### 예제 입력 1 
    UNUCIC
### 예제 출력 1 
    36

### Code

In [ ]:
userInput = list(input())
dialList = [list(i) for i in 'ABC DEF GHI JKL MNO PQRS TUV WXYZ'.split()]

totalDistance = 0

for alphabet in userInput:
  dialIndex = 2 + 1
  for dial in dialList:
    if alphabet in dial:
      dialIndex += dialList.index(dial)
    else:
      continue
  totalDistance += dialIndex

print(totalDistance)

36


## 2941. 크로아티아 알파벳

### 문제
    예전에는 운영체제에서 크로아티아 알파벳을 입력할 수가 없었다. 따라서, 다음과 같이 크로아티아 알파벳을 변경해서 입력했다.
    
    크로아티아 알파벳	>	변경
    č	>	c=
    ć	>	c-
    dž	>	dz=
    đ	>	d-
    lj	>	lj
    nj	>	nj
    š	>	s=
    ž	>	z=
    
    예를 들어, ljes=njak은 크로아티아 알파벳 6개(lj, e, š, nj, a, k)로 이루어져 있다. 단어가 주어졌을 때, 몇 개의 크로아티아 알파벳으로 이루어져 있는지 출력한다.
    
    dž는 무조건 하나의 알파벳으로 쓰이고, d와 ž가 분리된 것으로 보지 않는다. lj와 nj도 마찬가지이다. 위 목록에 없는 알파벳은 한 글자씩 센다.

### 입력
    첫째 줄에 최대 100글자의 단어가 주어진다. 알파벳 소문자와 '-', '='로만 이루어져 있다.
    
    단어는 크로아티아 알파벳으로 이루어져 있다. 문제 설명의 표에 나와있는 알파벳은 변경된 형태로 입력된다.

### 출력
    입력으로 주어진 단어가 몇 개의 크로아티아 알파벳으로 이루어져 있는지 출력한다.

### 예제 입력 1 
    ljes=njak
### 예제 출력 1 
    6
### 예제 입력 2 
    ddz=z=
### 예제 출력 2 
    3
### 예제 입력 3 
    nljj
### 예제 출력 3 
    3
### 예제 입력 4 
    c=c=
### 예제 출력 4 
    2

### Code

#### 1st try (**틀림**)

In [ ]:
userInput = input()
rWord = list(userInput[::-1])

croatianList = ['c=', 'c-', 'dz=', 'd-', 'lj', 'nj', 's=', 'z=']
numAlphabet = 0

for chr in rWord:
  isLast = rWord.index(chr) < len(rWord)-1
  if isLast:
    nextChr = rWord[rWord.index(chr)+1]
  else:
    nextChr = None
    
  if chr in ['=', '-', 'j'] and nextChr+chr in croatianList and isLast:
    if 
    rWord.remove(nextChr)
  else: pass

  rWord.remove(chr)
  numAlphabet += 1

print(numAlphabet)

ddz=z=
3


#### 2nd try

In [ ]:
userInput = input()
word = list(userInput)

croatianList = ['c=', 'c-', 'dz=', 'd-', 'lj', 'nj', 's=', 'z='] # 'dz=' is a invalid argument but existing for understanding
numCroatian = 0

while word:
  remaining = len(word)
  if remaining == 1:
    numCroatian += 1
    del word
    break
  elif remaining == 2:
    wordToCheck = ''.join(word[0:2])
    if wordToCheck in croatianList:
      del word[0:2]
    else:
      del word[0]
    numCroatian += 1
  else:
    wordToCheck = ''.join(word[0:2])
    if ''.join(word[0:3]) == 'dz=':
      del word[0:3]
    elif wordToCheck in croatianList:
      del word[0:2]
    else:
      del word[0]
    numCroatian += 1
    
print(numCroatian)

c=c=
2


## 1316. 그룹 단어 체커

### 문제
    그룹 단어란 단어에 존재하는 모든 문자에 대해서, 각 문자가 연속해서 나타나는 경우만을 말한다. 예를 들면, ccazzzzbb는 c, a, z, b가 모두 연속해서 나타나고, kin도 k, i, n이 연속해서 나타나기 때문에 그룹 단어이지만, aabbbccb는 b가 떨어져서 나타나기 때문에 그룹 단어가 아니다.

    단어 N개를 입력으로 받아 그룹 단어의 개수를 출력하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 단어의 개수 N이 들어온다. N은 100보다 작거나 같은 자연수이다. 둘째 줄부터 N개의 줄에 단어가 들어온다. 단어는 알파벳 소문자로만 되어있고 중복되지 않으며, 길이는 최대 100이다.

### 출력
    첫째 줄에 그룹 단어의 개수를 출력한다.

### 예제 입력 1 
    3
    happy
    new
    year
### 예제 출력 1 
    3
### 예제 입력 2 
    4
    aba
    abab
    abcabc
    a
### 예제 출력 2 
    1

### Code

In [ ]:
repeatNum = int(input())
numGroupWord = 0

for i in range(repeatNum):
  wordToCheck = input()
  charList = list(wordToCheck)
  charDict = {char:0 for char in charList}
  for char in charList:
    charDict[char] += 1
  while charList:
    firstChar = charList[0]
    charValue = charDict[firstChar]
    if ''.join(charList[0:charValue]) == firstChar*charValue:
      del charList[0:charValue]
    else:
      break
  else:
    numGroupWord += 1

print(numGroupWord)

4
aba
abab
abcabc
a
1


# 8. 기본 수학 1

수학 문제로 수학적 사고력을 길러 봅시다.

## 1712. 손익분기점

### 문제
    월드전자는 노트북을 제조하고 판매하는 회사이다. 노트북 판매 대수에 상관없이 매년 임대료, 재산세, 보험료, 급여 등 A만원의 고정 비용이 들며, 한 대의 노트북을 생산하는 데에는 재료비와 인건비 등 총 B만원의 가변 비용이 든다고 한다.

    예를 들어 A=1,000, B=70이라고 하자. 이 경우 노트북을 한 대 생산하는 데는 총 1,070만원이 들며, 열 대 생산하는 데는 총 1,700만원이 든다.

    노트북 가격이 C만원으로 책정되었다고 한다. 일반적으로 생산 대수를 늘려 가다 보면 어느 순간 총 수입(판매비용)이 총 비용(=고정비용+가변비용)보다 많아지게 된다. 최초로 총 수입이 총 비용보다 많아져 이익이 발생하는 지점을 손익분기점(BREAK-EVEN POINT)이라고 한다.

    A, B, C가 주어졌을 때, 손익분기점을 구하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 A, B, C가 빈 칸을 사이에 두고 순서대로 주어진다. A, B, C는 21억 이하의 자연수이다.

### 출력
    첫 번째 줄에 손익분기점 즉 최초로 이익이 발생하는 판매량을 출력한다. 손익분기점이 존재하지 않으면 -1을 출력한다.

### 예제 입력 1 
    1000 70 170
### 예제 출력 1 
    11
### 예제 입력 2 
    3 2 1
### 예제 출력 2 
    -1
### 노트
    이 문제에서 설명하는 손익분기점은 회계학에서 정의하는 손익분기점과 다를 수 있다.

### Code

#### 1st try

시간초과

In [ ]:
A, B, C = [int(i) for i in input().split(' ')]
numSales = 1

if B < C:
  while A+(B*numSales) >= C*numSales:
    numSales += 1
else:
  numSales = -1

print(numSales)

100 7 17
11


#### 2nd Try

In [ ]:
A, B, C = [int(i) for i in input().split(' ')]

if B < C:
  numSales = A//(C-B)+1
  print(numSales)
else:
  print(-1)

1000 70 170
11


## 2292. 벌집

#### 그림

https://www.acmicpc.net/JudgeOnline/upload/201009/3(2).png

### 문제
    위의 그림과 같이 육각형으로 이루어진 벌집이 있다. 그림에서 보는 바와 같이 중앙의 방 1부터 시작해서 이웃하는 방에 돌아가면서 1씩 증가하는 번호를 주소로 매길 수 있다. 숫자 N이 주어졌을 때, 벌집의 중앙 1에서 N번 방까지 최소 개수의 방을 지나서 갈 때 몇 개의 방을 지나가는지(시작과 끝을 포함하여)를 계산하는 프로그램을 작성하시오. 예를 들면, 13까지는 3개, 58까지는 5개를 지난다.

### 입력
    첫째 줄에 N(1 ≤ N ≤ 1,000,000,000)이 주어진다.

### 출력
    입력으로 주어진 방까지 최소 개수의 방을 지나서 갈 때 몇 개의 방을 지나는지 출력한다.

### 예제 입력 1 
    13
### 예제 출력 1 
    3

### Code

#### 1st try

오답. 채점 중간에 틀림

In [ ]:
number = int(input())-1
step = 1

while number >= (6*step):
  number -= (6*step)
  step += 1
else:
  step += 1

print(step)

70
6


#### 2nd try

In [ ]:
number = int(input())-1
layer = 1

while number > 0:
  number -= layer*6
  layer += 1

print(layer)

58
5


## 1193. 분수찾기

### 문제
    무한히 큰 배열에 다음과 같이 분수들이 적혀있다.

1/1	1/2	1/3	1/4	1/5	…
2/1	2/2	2/3	2/4	…	…
3/1	3/2	3/3	…	…	…
4/1	4/2	…	…	…	…
5/1	…	…	…	…	…
…	…	…	…	…	…

    이와 같이 나열된 분수들을 1/1 -> 1/2 -> 2/1 -> 3/1 -> 2/2 -> … 과 같은 지그재그 순서로 차례대로 1번, 2번, 3번, 4번, 5번, … 분수라고 하자.

    X가 주어졌을 때, X번째 분수를 구하는 프로그램을 작성하시오.

### 입력
    첫째 줄에 X(1 ≤ X ≤ 10,000,000)가 주어진다.

### 출력
    첫째 줄에 분수를 출력한다.

### 예제 입력 1 
    14
### 예제 출력 1 
    2/4

### Code

In [ ]:
number = int(input())
x = y = layer = 1

while layer < number:
  number -= layer
  layer += 1

if layer%2 == 1:
  x, y = layer, 1
  x -= (number-1)
  y += (number-1)
else:
  x, y = 1, layer
  x += (number-1)
  y -= (number-1)
  
print(f'{x}/{y}')

15
1/5
